### PR TITLE
stampit({options}) api

### PR DESCRIPTION
**Pull request by koresar**
Please, review. List of changes:
* New `stampit( options )` API.
* Convenience shortcut methods, like `stampit.methods()`, etc. See #41.
* Reorganized tests a bit for more convenience.
* Drop IE8 support.
* `/dist/` is now not committed to the repo but build only on the `prepublish` NPM hook.
* Excluded `/test/`, travis, jshist, SauceLabs, etc. development files from NPM package via `.npmignore`.

### CODE REVIEW

**Comment by ericelliott** [comment]
> * /dist/ is now not committed to the repo but build only on the prepublish NPM hook.

Do you have a link to a reference where I can learn more about this? Is the latest `/dist/` build still available when people run `npm install` in their apps? IMO, it's a little weird that it won't be browsable in the repository on GitHub...

> * Excluded /test/, travis, jshist, SauceLabs, etc. development files from NPM package via .npmignore.

Probably a good idea.

File: stampit.js [hunk 1]
```diff
@@ -92,10 +91,7 @@ function compose(factories) {
  * Return a factory function that will produce new objects using the
  * prototypes that are passed in or composed.
  *
- * @param  {Object} [methods] A map of method names and bodies for delegation.
- * @param  {Object} [refs]   A map of property names and values to be mixed into each new object.
- * @param  {Function} [init] A closure (function) used to create private data and privileged methods.
- * @param  {Object} [props]   An object to be deeply cloned into each newly stamped object.
+ * @param  {Object} [options] Options to build stamp from: `{ methods, refs, init, props }`
```
Comment (ericelliott) [inline stampit.js]: There should still be `@param` annotations for each of:

`options.methods`, `options.refs`, `options.init`, and `options.props`.

I wonder if it's possible to define an options type with JSDoc without it assuming that there's an `options` class?

Then you could define the type elsewhere and just do `@param {options} [options] desc...`

**Comment by troutowicz** [comment]
@ericelliott 
>Do you have a link to a reference where I can learn more about this? Is the latest /dist/ build still available when people run npm install in their apps? 

https://docs.npmjs.com/misc/scripts

First bullet point. There's really no need to have dist/ in the repo, noone looks at it. It's just for compatibility.

@koresar We should use the `files` key of package.json instead of `.npmignore`.
https://docs.npmjs.com/files/package.json#files

```js
files: {
  stampit.js,
  mixer.js,
  ... other files to include
}
```

File: test/index.js [hunk 2]
```diff
@@ -1,11 +1 @@
...
-require('./convert-constructor');
-require('./init');
-require('./immutability');
-require('./is-stamp');
-require('./props-safety');
-require('./refs');
-require('./stampit-aliases');
-require('./state');
-require('./v1-compatibility');
+require('require-all')(__dirname);
```
Comment (ericelliott) [inline test/index.js]: :+1:

File: test/stampit-api.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  t.end();
+});
+
+test('stampit({})', function (t) {
+  t.ok(stampit.isStamp(stampit({})));
+
+  t.end();
+});
+
+test('incorrect stampit({ methods }) args', function (t) {
```
Comment (ericelliott) [inline test/stampit-api.js]: Good thinking. Just to clarify, these all result in stamps with no methods, right?

File: test/stampit-shortcuts.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
+'use strict';
+var stampit = require('../stampit'),
+  test = require('tape');
+
+// stampit.methods, stampit.refs, stampit.init, stampit.props
+
+test('stampit.methods shortcut', function (t) {
```
Comment (ericelliott) [inline test/stampit-shortcuts.js]: Nice!

File: test/stampit-shortcuts.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  var init = { method1: function(){} };
+  var stamp1 = stampit({ init: init });
+  var stamp2 = stampit.init(init);
+
+  t.deepEqual(stamp1.fixed, stamp2.fixed);
+
+  t.end();
+});
+
+test('stampit.props shortcut', function (t) {
```
Comment (ericelliott) [inline test/stampit-shortcuts.js]: w00t. Shortcuts rock. :+1:

**Comment by ericelliott** [comment]
@troutowicz Thanks for the reference. I was unaware of this part:

> "(Also run on local npm install without any arguments.)"

I think that answers my question, so the dist build of stampit will be available if we add it as `"main"`, right?

I ask because I'd love to convert Stampit source to ES6 and compile it, but we'll still need to support require in ES5 projects. =)

**Comment by troutowicz** [comment]
Exactly.

```js
"main": "dist/stampit.js"
```

I can help out with the ES6 conversion. I recently submitted a PR to @koresar's supermixer repo doing just that. I want to see all stampit projects using ES6!

**Comment by ericelliott** [comment]
> I want to see all stampit projects using ES6!

:+1:

**Comment by ericelliott** [comment]
My comments are just minor nitpicks. I'm OK with landing this.

1. Are there other PRs we need to merge first?
2. Are the rest of the to-do items for 2.0 ready?

Please feel free to merge any pending PRs necessary to land 2.0. I'm excited to share all your great contributions with the world. =)

File: test/stampit-api.js (hunk 3 above)
Comment (koresar) [inline test/stampit-api.js]: Yeah. More checks should be added. Thanks for pointing out.

File: test/stampit-shortcuts.js (hunk 4 above)
Comment (koresar) [inline test/stampit-shortcuts.js]: Cheers :)

File: stampit.js (hunk 1 above)
Comment (koresar) [inline stampit.js]: There is a way. I'll see what I can do.

**Comment by koresar** [comment]
@ericelliott 
1. There are no outstanding PR's to my knowledge.
2. I'm trying to keep the outstanding things in the [Milestone 2.0](https://github.com/ericelliott/stampit/milestones/2.0). The only thing left is README.

RE the `prepublish` and `dist/`. Each time you'll execute `npm publish` the `dist/` will be generated and uploaded to the NPM registry.

**Comment by koresar** [comment]
@troutowicz I do not feel strong about `files` feature in `package.json`. I'll leave `.npmignore` for the time being. We'll see how it goes.

**Comment by koresar** [comment]
Updated the JSDoc of `stampit()` to describe argument. JSDoc docs: http://usejsdoc.org/tags-param.html

**Comment by koresar** [comment]
#74 got implemented as well (static stuff). Requires thorough review. @ericelliott ?

**Comment by ericelliott** [comment]
[The io.js install is failing](https://travis-ci.org/ericelliott/stampit/jobs/64796670).

Looks like Travis can't find lodash 3.9.1+ even though it's currently at 3.9.3. Am I reading that right?

**Comment by koresar** [comment]
Correct. I have no idea why is that happening. I tried to use earlier lodash, but it fails with "Can't download package from the HTTPS uri". Might be an NPM issue (too old build).

**Comment by ericelliott** [comment]
This looks good to me. I'm happy to merge it but maybe we should hold off on npm publish until we sort out why the travis build is failing?

**Comment by koresar** [comment]
You can fixing it. :) Just put the following into the travis yaml.
```yaml
before_install
  npm update npm -g
```

**Comment by ericelliott** [comment]
Oh, cool! Can you do that in a new PR? This one is already merged. =)

**Comment by koresar** [comment]
Btw, Travis-CI fixed their `io.js`+`lodash` bug. Builds are now green. :)

**Comment by ericelliott** [comment]
Nice.

**Pull request merged**